garrysmod/lua/weapons/weapon_fists.lua

```diff
@@ -18,216 +18,303 @@ SWEP.UseHands = true
+	if ( !owner:IsValid() ) then return end
 
-end
+	local tr = self:PunchTrace()
+	local hitent = tr.Entity
 
-function SWEP:Deploy()
+	if ( !self:CanHit( hitent ) ) then
+		self:SetCombo( 0 )
+		return
+	end
 
-	local speed = GetConVarNumber( "sv_defaultdeployspeed" )
+	self:EmitSound( self.HitSound )
 
-	local vm = self:GetOwner():GetViewModel()
-	vm:SendViewModelMatchingSequence( vm:LookupSequence( "fists_draw" ) )
-	vm:SetPlaybackRate( speed )
+	local phys = hitent:GetPhysicsObject()
+	local physscale = phys_pushscale:GetFloat()
+	local hitpos = tr.HitPos
 
-	self:SetNextPrimaryFire( CurTime() + vm:SequenceDuration() / speed )
-	self:SetNextSecondaryFire( CurTime() + vm:SequenceDuration() / speed )
-	self:UpdateNextIdle()
+	if ( phys:IsValid() ) then
+		-- FIXME: This cannot be right
+		phys:ApplyForceOffset( aimvector * phys:GetMass() * physscale * self.PunchForceScale, hitpos )
+	end
 
-	if ( SERVER ) then
+	if ( !self:CanDamage( hitent ) ) then
 		self:SetCombo( 0 )
+		return
 	end
 
-	return true
-
-end
+	local dmginfo = DamageInfo()
+	dmginfo:SetAttacker( owner )
+	dmginfo:SetInflictor( self )
+	dmginfo:SetDamagePosition( hitpos )
+	dmginfo:SetDamageType( DMG_CLUB )
+	dmginfo:SetReportedPosition( tr.StartPos )
 
-function SWEP:Holster()
+	local dmg, force
+	local combo = self:GetCombo()
 
-	self:SetNextMeleeAttack( 0 )
+	if ( combo >= self.ComboCount ) then
+		dmg = self.ComboDamage
+		force = self.ComboForce
+		self:SetCombo( 0 )
+	else
+		dmg = self.PunchDamage
+		force = self.PunchForce
+		self:SetCombo( combo + 1 )
+	end
 
-	return true
+	dmginfo:SetDamage( util.SharedRandom( "GMOD_" .. self:GetClass() .. self:EntIndex(), dmg.Min, dmg.Max ) )
+	dmginfo:SetDamageForce(
+		owner:GetForward() * force.Forward * physscale +
+		owner:GetRight() * force.Right * physscale * ( self:GetRightPunch() && -1 || 1 ) +
+		owner:GetUp() * force.Up * physscale )
 
+	-- FIXME
+	SuppressHostEvents( NULL ) -- Let the breakable gibs spawn in multiplayer on client
+	hitent:DispatchTraceAttack( dmginfo, tr )
+	SuppressHostEvents( owner )
 end
 
-function SWEP:Think()
+function SWEP:OnDrop()
 
-	local vm = self:GetOwner():GetViewModel()
-	local curtime = CurTime()
-	local idletime = self:GetNextIdle()
+	self:Remove() -- You can't drop fists
 
-	if ( idletime > 0 && CurTime() > idletime ) then
+end
 
-		vm:SendViewModelMatchingSequence( vm:LookupSequence( "fists_idle_0" .. math.random( 1, 2 ) ) )
+function SWEP:Holster()
 
-		self:UpdateNextIdle()
+	self:SetCombo( 0 )
+	self:SetPunchHitTime( 0 )
 
-	end
+	return true
 
-	local meleetime = self:GetNextMeleeAttack()
+end
 
-	if ( meleetime > 0 && CurTime() > meleetime ) then
+-- Punching should only be done on the server in single-player
+if ( CLIENT and game.SinglePlayer() ) then return end
 
-		self:DealDamage()
+function SWEP:Think()
 
-		self:SetNextMeleeAttack( 0 )
+	local curtime = CurTime()
+	local punchhit = self:GetPunchHitTime()
 
+	if ( punchhit > 0 && punchhit <= curtime ) then

```

Comment:
Inconsistent usage of `&&` and `and` within the same file. Should just be `and`. There are other cases of this.